IdentityCore/src/validation/MSIDAuthority.m

```diff
@@ -137,4 +197,145 @@ + (NSURL *)cacheUrlForAuthority:(NSURL *)authority
+        resolver = [MSIDB2CAuthorityResolver new];
+    }
+    // AAD.
+    else
+    {
+        resolver = [MSIDAadAuthorityResolver new];
+    }
+    
+    [resolver discoverAuthority:authority
+              userPrincipalName:upn
+                       validate:validate
+                        context:context
+                completionBlock:completionBlock];
+}
+
++ (void)loadOpenIdConfigurationInfo:(NSURL *)openIdConfigurationEndpoint
+                            context:(id<MSIDRequestContext>)context
+                    completionBlock:(MSIDOpenIdConfigurationInfoBlock)completionBlock
+{
+    __auto_type cacheKey = openIdConfigurationEndpoint.absoluteString.lowercaseString;
+    __auto_type metadata = [s_openIdConfigurationCache objectForKey:cacheKey];
+    
+    if (metadata)
+    {
+        if (completionBlock) completionBlock(metadata, nil);
+        return;
+    }
+    
+    __auto_type request = [[MSIDOpenIdConfigurationInfoRequest alloc] initWithEndpoint:openIdConfigurationEndpoint];
+    [request sendWithBlock:^(MSIDOpenIdProviderMetadata *metadata, NSError *error)
+    {
+        if (cacheKey && metadata)
+        {
+            [s_openIdConfigurationCache setObject:metadata forKey:cacheKey];
+        }
+        
+        if (completionBlock) completionBlock(metadata, error);
+    }];
+}
+
++ (NSURL *)normalizeAuthority:(NSURL *)authority
+                      context:(id<MSIDRequestContext>)context
+                        error:(NSError **)error
+{
+    if (![MSIDAuthority isAuthorityFormatValid:authority context:context error:error])
+    {
+        return nil;
+    }
+    
+    // B2C
+    if ([self isB2CInstanceURL:authority])
+    {
+        NSString *updatedAuthorityString = [NSString stringWithFormat:@"https://%@/%@/%@/%@", [authority msidHostWithPortIfNecessary], authority.pathComponents[1], authority.pathComponents[2], authority.pathComponents[3]];
+        return [NSURL URLWithString:updatedAuthorityString];
+    }
+    
+    // ADFS and AAD
+    return [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [authority msidHostWithPortIfNecessary], authority.pathComponents[1]]];
+}
+
++ (BOOL)isAuthorityFormatValid:(NSURL *)authority
+                       context:(id<MSIDRequestContext>)context
+                         error:(NSError **)error
+{
+    if ([NSString msidIsStringNilOrBlank:authority.absoluteString])
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"'authority' is a required parameter and must not be nil or empty.", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    if (![authority.scheme isEqualToString:@"https"])
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"authority must use HTTPS.", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    if (authority.pathComponents.count < 2)
```

Comment:
again, this is realistic for AAD, but for other authorities there might be less than 2 components, because there's no tenant. Can we make sure we separate the "base" implementation from AAD specific things.